### PR TITLE
Add support for MacOS

### PR DESCRIPTION
**Pull request by proski**
This is for x86_64 only. No attempt was made to support Apple Silicon. However, this PR could be used as a spepping stone in that direction.

The remaining problem is an error when running `prpll`:

```
Exception gpu_error: INVALID_WORK_GROUP_SIZE (-54) transpIn at /Users/plroskin/src/gpuowl/src/clwrap.cpp:312 run
```

That would need to be addressed by an OpenCL expert. It's possible that the issue is specific to my hardware or software.

However, many other issues have been addressed is a clean way.

* The real gcc (e.g. from homebrew) is selected for the build (clang doesn't support 128-bit floating point and `std::filesystem::path` in the standard library).
* OpenCL is linked the way is should be on MacOS with the `-framework OpenCL` linker directive.
* The cmake build uses cmake facilities to search for OpenCL and link to it.
* A replacement for `fdatasync()` is used on MacOS.
* OpenCL functions have prototypes when needed.
* Comparison between `i32` and `u32` in OpenCL code has been fixed.
* `transposeWords` uses signature that preserves `global` and `restrict` attributes used by the callers.
* `clCreateCommandQueueWithProperties` is missing on MacOS, use `clCreateCommandQueueWithPropertiesAPPLE` instead.

Basically, the C++ code compiles, the OpenCL code compiles and loads, but fails to run on the GPU.

### CODE REVIEW

**Comment by tdulcet** [comment]
@proski - If you want future commits to be automatically checked that they successfully build on x86 macOS, you could add something like this to the existing GitHub actions [workflow file](https://github.com/preda/gpuowl/blob/master/.github/workflows/ci.yml):
```yml
  macOS:
    name: macOS

    runs-on: macos-13
    steps:
    - uses: actions/checkout@v4
    - name: Install
      run: |
        brew install opencl-headers
        g++ --version
    - name: Script
      run: |
        make prpll -j "$(sysctl -n hw.ncpu)"
        cd build-release
        rm -f -- *.o
        ./prpll -h
    - uses: actions/upload-artifact@v4
      if: always()
      with:
        name: macos_prpll
        path: ${{ github.workspace }}
```
I am guessing that the [opencl-headers](https://formulae.brew.sh/formula/opencl-headers) package would need to be installed. The [opencl-icd-loader](https://formulae.brew.sh/formula/opencl-icd-loader) package might need to be installed as well.

**Comment by proski** [comment]
> @proski - If you want future commits to be automatically checked that they successfully build on x86 macOS, you could add something like this to the existing GitHub actions [workflow file](https://github.com/preda/gpuowl/blob/master/.github/workflows/ci.yml):

I've added it, thank you!

> I am guessing that the [opencl-headers](https://formulae.brew.sh/formula/opencl-headers) package would need to be installed. The [opencl-icd-loader](https://formulae.brew.sh/formula/opencl-icd-loader) package might need to be installed as well.

None of them is needed. Even `gcc@14` is installed already, but I'm installing it anyway to be explicit about the dependency.

We should run a real test like `prpll -iters 10000 -ll 82589933` but other OSes don't do it, so let's keep it consistent. Also, there is no test for the cmake build, I should have added it in #293. It could be added in a separate PR.

**Comment by proski** [comment]
`INVALID_WORK_GROUP_SIZE` is reproducible in the CI:

https://github.com/preda/gpuowl/actions/runs/10842867463/job/30089143591?pr=295#step:4:71

File: src/cl/base.cl [hunk 1]
```diff
@@ -138,12 +138,17 @@ typedef global const double2* BigTab;
 
 #define KERNEL(x) kernel __attribute__((reqd_work_group_size(x, 1, 1))) void
 
+// Prototypes
+void read(u32 WG, u32 N, T2 *u, const global T2 *in, u32 base);
```
Comment (preda) [inline src/cl/base.cl:142]: Interesting, are these prototypes needed? does the compiler produce some warnings without them?

**Review by preda — APPROVED** [review]

File: src/cl/base.cl (hunk 1 above)
Comment (proski) [reply src/cl/base.cl:142]: Yes, they fix warnings. Presumably, the OpenCL compiler on MacOS is unnecessarily strict.

```
./build-release/prpll -ll 82589933
20240913 23:12:47  PRPLL 0.15-4-g3a1361d-dirty starting
20240913 23:12:47  config: -ll 82589933
20240913 23:12:47  device 0, OpenCL 1.1, unique id ''
20240913 23:12:47 82589933 config:
20240913 23:12:47 82589933 FFT: 4.50M 1K:9:256:1:0 (17.50 bpw)
20240913 23:12:47 82589933 In file included from <program source>:1:
In file included from transpose.cl:4:
base.cl:147:6: warning: no previous prototype for function 'read'
void read(u32 WG, u32 N, T2 *u, const global T2 *in, u32 base) {
     ^

20240913 23:12:47 82589933 In file included from <program source>:1:
In file included from transpose.cl:4:
base.cl:147:6: warning: no previous prototype for function 'read'
void read(u32 WG, u32 N, T2 *u, const global T2 *in, u32 base) {
     ^

20240913 23:12:47  Exception gpu_error: INVALID_WORK_GROUP_SIZE (-54) transpIn at src/clwrap.cpp:312 run
20240913 23:12:47  Bye
```
